### ISSUE TITLE
Loading of 20 million records takes 6 minutes

### ISSUE THREAD

**Issue by ElijahGM** [octobercms/october#2788]
##### Expected behavior
I have a list of 20 million records i want it to load faster in milliseconds and allow filtering (in list time possible) and export reords based on filters
##### Actual behavior
Trying to load the List it is taking close to 6 minutes to load the  page, when i try to to export i get a 500 error

##### Reproduce steps
Click on controller index action to list all records (Loads after 6 minutes)
Click on Export (Application crushes)

##### October build
(October build number)

**Comment by daftspunk** [comment]
This is more of a DevOps issue. I assume the error 500 is caused by a lack of server resources to process the action. Again with the 6 minute load time. How much RAM do you have allocated for the server and for the PHP and MySQL processes? The bottleneck could be anywhere really. Consult your server administrator to pinpoint exactly where this is.

**Comment by ElijahGM** [comment]
I have 32GB of ram , when I load 3k records it works perfectly, the problem
is when I have 20m records,
For the error 500, when I restart the server it resumes perfectly, when I
try to export again, takes time then when I refresh the page I get
permanent error 500, until I restart again, all other application works
well throughout, when I try to create a Php raw or yii2 application with
the same records they work well.
 Please advice

On Mar 31, 2017 7:34 AM, "Samuel Georges" <notifications@github.com> wrote:

This is more of a DevOps issue. I assume the error 500 is caused by a lack
of server resources to process the action. Again with the 6 minute load
time. How much RAM do you have allocated for the server and for the PHP and
MySQL processes? The bottleneck could be anywhere really. Consult your
server administrator to pinpoint exactly where this is.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/octobercms/october/issues/2788#issuecomment-290613632>,
or mute the thread

**Comment by daftspunk** [comment]
Ok, we need someone to benchmark the ListController with 20m records and tell us where the bottleneck is. Sounds like it could be a fun exercise for someone.

**Comment by LukeTowers** [comment]
With that many records, wouldn't it be better to use `DB::raw()` statements instead of fully instantiated Collection and Model objects for all the records?

**Comment by daftspunk** [comment]
@LukeTowers I don't think it would matter too much because the Collection would only contain the number of records shown on a given page. Unless pagination is disabled, this wouldn't be a cause for concern.

**Comment by LukeTowers** [comment]
```
Click on controller index action to list all records (Loads after 6 minutes)
Click on Export (Application crushes)
```
^ Pagination is disabled in those cases, is it not?

**Comment by ElijahGM** [comment]
How do i enable pagination on both cases,
For Export can we have a queue feature where if records are many the request is queued for spees and performance?

**Comment by daftspunk** [comment]
> ^ Pagination is disabled in those cases, is it not?

Right, that'll be the problem, please use pagination on the ListController and implement [chunking](https://octobercms.com/docs/database/query#retrieving-results) in your export model. 

Without Pagination you'll need to implement a custom rendering solution that uses the lower level `Db` class to perform queries, this is far more efficient and will give the same level of performance as other frameworks, such as Yii.

Alternatively for exporting you could push a task on to the queue to compile the export file then send you an email when it has finished.

**Comment by LukeTowers** [comment]
@ElijahGM to setup pagination on ListControllers, use the `recordsPerPage` property in your `config_list.yaml` file. http://octobercms.com/docs/backend/lists#configuring-list

**Comment by ElijahGM** [comment]
I think the time is taken creating paginations

**Comment by LukeTowers** [comment]
We're going to need more information to be able to figure this out.

Have you enabled pagination on your list controller? What actions exactly are you taking that demonstrate the slowdown?

**Comment by ElijahGM** [comment]
It seems the js script that is creating the list on the frontend is what is taking time

**Comment by LukeTowers** [comment]
That doesn't sound right, the list controller should be chunking the records on the backend before serving them up as pages to the frontend, so the JS script shouldn't have to be paging 20 million records. Could you check if that is actually the case or if we need to make some improvements to the list controller behaviour?

**Comment by ElijahGM** [comment]
The list Controller is chunking correctly but on rendering the scripts
takes time to render the pagination widget until the browser mark its as
non-responsive script that need to be terminated

**Comment by LukeTowers** [comment]
Are you able to look into that script and let us know what's causing the delay? Is it the part of the script that renders the page selector (with 20 million records, and a default chunk of 20 per page, that means that the page selector could potentially be trying to show a million options). If that's the case, then we need to revisit the implementation of the page selector to have a more dynamic solution for that then rendering a separate DOM element for each page option that you could jump to.

**Comment by ElijahGM** [comment]
I thing the problem is that the dropdown selector is loading 20 million
records

**Comment by LukeTowers** [comment]
It wouldn't be 20 million, it would be 1 million, because it generates I'm guessing that it generates one `<option>` element per page (20 records by default) of results. Are you able to look into more dynamic ways that could be done and make a PR?

**Comment by ElijahGM** [comment]
Let me see what i can do and let you know

**Comment by jofrysutanto** [comment]
We stumble on the same issue where we have 12 million records in our database, and using the default ListController setting with record-per-page setting of 15 causing the browser to quickly become unresponsive due to the attempt of rendering 800,000 page dropdown.

This might means that there is real benefit of allowing the switch between LengthAwarePaginator and the simple Paginator.

For example:

```
# config_list.yaml

# Records to display per page
recordsPerPage: 20

# Show page number
# Setting this to false will hide page number dropdown selection,
# this is suitable for large tables
showPageNumber: true
```

We then make changes to `ListController` and the `Lists` widget to accept this additional configuration, enough so that in `Lists@getRecords`:

```
    /**
     * Returns all the records from the supplied model, after filtering.
     * @return Collection
     */
    protected function getRecords()
    {
        $model = $this->prepareModel();

        if ($this->showTree) {
            $records = $model->getNested();
        }
        elseif ($this->showPagination) {
            $records = $this->showPageNumber 
                ? $model->paginate($this->recordsPerPage, $this->currentPageNumber)
                : $model->simplePaginate($this->recordsPerPage, $this->currentPageNumber);
        }
        else {
            $records = $model->get();
        }

        /*
         * Extensibility
         */
        if ($event = $this->fireSystemEvent('backend.list.extendRecords', [&$records])) {
            $records = $event;
        }

        return $this->records = $records;
    }
```

We will also need to modify the template that renders the pagination, and there might be other consideration in place which I am not aware of, but hopefully this illustrates clearly how we could improve the List.

**Comment by LukeTowers** [comment]
That proposal sounds good to me @jofrysutanto, open a PR for it :)

**Comment by LukeTowers** [comment]
@ElijahGM this has been fixed by #3084 with the contributions from @jofrysutanto.
